source/extensions/filters/http/common/compressor/compressor.cc

```diff
@@ -0,0 +1,315 @@
+                       : StringUtil::CaseUnorderedSet(types.cbegin(), types.cend());
+}
+
+uint32_t CompressorFilterConfig::contentLengthUint(Protobuf::uint32 length) {
+  return length > 0 ? length : DefaultMinimumContentLength;
+}
+
+CompressorFilter::CompressorFilter(CompressorFilterConfigSharedPtr config)
+    : skip_compression_{true}, compressor_(), config_(std::move(config)) {}
+
+Http::FilterHeadersStatus CompressorFilter::decodeHeaders(Http::HeaderMap& headers, bool) {
+  if (config_->runtime().snapshot().featureEnabled(config_->featureName(), 100) &&
+      isAcceptEncodingAllowed(headers)) {
+    compressor_ = config_->makeCompressor();
+    skip_compression_ = false;
+    if (config_->removeAcceptEncodingHeader()) {
+      headers.removeAcceptEncoding();
+    }
+  } else {
+    config_->stats().not_compressed_.inc();
+  }
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterHeadersStatus CompressorFilter::encodeHeaders(Http::HeaderMap& headers,
+                                                          bool end_stream) {
+  if (!end_stream && !skip_compression_ && isMinimumContentLength(headers) &&
+      isContentTypeAllowed(headers) && !hasCacheControlNoTransform(headers) &&
+      isEtagAllowed(headers) && isTransferEncodingAllowed(headers) && !headers.ContentEncoding()) {
+    sanitizeEtagHeader(headers);
+    insertVaryHeader(headers);
+    headers.removeContentLength();
+    headers.insertContentEncoding().value(config_->contentEncoding());
+    config_->stats().compressed_.inc();
+  } else if (!skip_compression_) {
+    skip_compression_ = true;
+    config_->stats().not_compressed_.inc();
+  }
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterDataStatus CompressorFilter::encodeData(Buffer::Instance& data, bool end_stream) {
+  if (!skip_compression_) {
+    config_->stats().total_uncompressed_bytes_.add(data.length());
+    compressor_->compress(data, end_stream ? Compressor::State::Finish : Compressor::State::Flush);
+    config_->stats().total_compressed_bytes_.add(data.length());
+  }
+  return Http::FilterDataStatus::Continue;
+}
+
+Http::FilterTrailersStatus CompressorFilter::encodeTrailers(Http::HeaderMap&) {
+  if (!skip_compression_) {
+    Buffer::OwnedImpl empty_buffer;
+    compressor_->compress(empty_buffer, Compressor::State::Finish);
+    config_->stats().total_compressed_bytes_.add(empty_buffer.length());
+    encoder_callbacks_->addEncodedData(empty_buffer, true);
+  }
+  return Http::FilterTrailersStatus::Continue;
+}
+
+bool CompressorFilter::hasCacheControlNoTransform(Http::HeaderMap& headers) const {
+  const Http::HeaderEntry* cache_control = headers.CacheControl();
+  if (cache_control) {
+    return StringUtil::caseFindToken(cache_control->value().getStringView(), ",",
+                                     Http::Headers::get().CacheControlValues.NoTransform);
+  }
+
+  return false;
+}
+
+bool CompressorFilter::isAcceptEncodingAllowed(const Http::HeaderMap& headers) const {
```

Comment:
this somewhat complex impl is figuring out which compressor to use but then dropping that info and returning a bool.

is the logic then repeated somewhere else? If so, could we implement this bool method as:
```
  return findCompressor(headers) != nullptr;
```
or similar?

Reply:
As far as I know this logic is relevant to compression filters only and every compression filter instantiates one `CompreesionFilter`. Even though the method returns bool it also updates separate stats for every filter it's called from.

Reply:
I'm missing something about how the system works. I feel like we should be computing operational knowledge of the compression choice once per request, and it looks like we are computing it and dropping it on the floor, keeping only whether or not it should be compressed.

Don't we need to compute which one to use and then actually run the compressor, as well? Or at least record in the request somehow an enum indicating compressor choice or something?

Reply:
See https://github.com/envoyproxy/envoy/blob/master/include/envoy/stream_info/filter_state.h for another option which seems a bit cleaner (strongly typed, no possible leakage onto the wire)